pkg/retry/retry.go

```diff
@@ -81,6 +82,13 @@ func IsErrorRetryable(err error) bool {
 			return false
 		}
 		return true
+	case *docker.UnexpectedHTTPStatusError:
+		// Retry on 5xx http server errors, they appear to be quite common in the field.
+		// https://github.com/containers/common/issues/2299
+		if e.StatusCode/100 == 5 {
```

Comment:
- 501 Not Implemented ?
- 505 HTTP Version Not Supported ?
- 507 Insufficient Storage (“If the request that received this status code was the result of a user action, the request MUST NOT be repeated until it is requested by a separate user action.”)?

This is not ultimately blocking: I’m generally skeptical about blind retries, and worried that they might hurt more than they help. But, also, I would have originally objected to retrying on 502 as insufficiently explicit about retries being valuable, so I wouldn’t trust my own judgement :)

I didn’t now carefully evaluate each individual status code, let’s first discuss whether we want to deal with things at  that level.

Reply:
Fair points, I think we can start small and slowly add more codes if we see the need.

502 is the very common one with quay.io that happens all the time. I guess that is just a load balancer retuning it as the backend server went unavailable. I assume that retries would bring you do a different backend and thus commonly work.
I have seen other codes from quay.io but I don't have the logs anymore to know the exact details. From memory I think they were 503 and 504.

I am fine to start with 502 and then we can reconsider others as we encounter them.

Reply:
Personally, I recommend checking 5xx errors that tend to be reported by reverse proxies/load balancers if the underlying service is unavailable:

- 502 Bad Gateway
- 503 Service Unavailable
- 504 Gateway Timeout

Reply:
I didn’t actually _go to read the specs_, but intuitively that sounds reasonable to me.

Reply:
… and if we are retrying on the catch-all “something behind a proxy failed” 502, I guess retrying on bare “something failed without a proxy” 500 would be consistent?

It _is_ the case that some hard failures (“unsupported manifest format”) show up as 500… but would that be 500 or 502 when fronted by a reverse proxy / load balancer?